solutions/IOI 13-game.mdx

```diff
@@ -0,0 +1,337 @@
+---
+id: ioi-13-game
+source: IOI 2013
+title: Game
+author: Andi Qu
+---
+
+**Time Complexity:** $\mathcal O(Q \log R \cdot \log C)$
+
+We're asked to handle point updates and range GCD queries on a 2D grid.
+This implies that we should use a 2D range-query data structure like a 2D
+segment tree (N.B. not a Fenwick tree, as the GCD function is not commutative).
+
+In 1D ($C = 1$), this can be solved by a fairly straightforward use of a segment tree:
+each node stores the GCD of its two children. Since $R$ can be quite big, this
+needs to be a sparse segment tree; another alternative would be a balanced binary
+tree like a treap.
+
+However, a sparse 2D segment tree uses *just* a bit too much memory, and only 
+scores 80 points. Fortunately for us, there are two ways to get around this!
+
+## Approach 1 - Sparse segment tree of BBSTs
+
+Although BBSTs use 4 times less memory than segment trees, a BBST of BBSTs is
```

Comment:
> a BBST of BBSTs is not possible

er actually range trees are BBSTs of BBSTs https://en.wikipedia.org/wiki/Range_tree

if the tree gets too unbalanced after a certain # of insertions then you end up rebuilding an entire subtree I think